# 1. Importing the Data

In [51]:
import sys
sys.path.append('../../src')

In [1]:
import requests
import pandas as pd
import io

# Your file id (replace with your actual file_id)
file_id = "1wx5xARwV83ejIzFi8mT9n3IjYK45fNm1"

# First request to get the confirmation prompt page
url = f"https://drive.google.com/uc?export=download&id={file_id}"
response = requests.get(url)

# In some cases, you might need to pass a confirmation token to download the file
# This depends on whether Google Drive presents a confirmation screen
confirm = None
for line in response.text.split('\n'):
    if "confirm=" in line:
        # Extract the confirmation token
        confirm = line.split('confirm=')[1].split('&')[0]
        break

# If we have a confirmation token, we need to append it to the URL
if confirm:
    url = f"https://drive.google.com/uc?export=download&id={file_id}&confirm={confirm}"
    response = requests.get(url)

# Check if the request was successful
if response.status_code == 200:
    # The output file name
    output = 'data_cleaned.pkl.gz'
    # Write the response content to a file (this will handle the gzip compression automatically)
    with open(output, "wb") as f:
        f.write(response.content)
    
    # Now, read the pickle file as a pandas DataFrame
    df = pd.read_pickle(output)
else:
    print("Failed to retrieve the data")


In [2]:
df.head()

searchDate  flightDate startingAirport destinationAirport  isBasicEconomy  \
0  2022-04-16  2022-04-17             ATL                BOS           False   
1  2022-04-16  2022-04-17             ATL                BOS           False   
2  2022-04-16  2022-04-17             ATL                BOS           False   
3  2022-04-16  2022-04-17             ATL                BOS           False   
4  2022-04-16  2022-04-17             ATL                BOS           False   

   isRefundable  isNonStop   totalFare  totalTravelDistance  \
0         False       True  248.600006                947.0   
1         False       True  248.600006                947.0   
2         False       True  248.600006                947.0   
3         False       True  248.600006                947.0   
4         False       True  248.600006                947.0   

   segmentsDepartureTimeEpochSeconds_Leg1  ...  \
0                            1.650215e+09  ...   
1                            1.650191e+09  ...   
2                            1.650210e+09  ...   
3                            1.650218e+09  ...   
4                            1.650204e+09  ...   

   segmentsDurationInSeconds_Leg4  segmentsDistance_Leg1  \
0                             0.0                  947.0   
1                             0.0                  947.0   
2                             0.0                  947.0   
3                             0.0                  947.0   
4                             0.0                  947.0   

   segmentsDistance_Leg2 segmentsDistance_Leg3 segmentsDistance_Leg4  \
0                    0.0                   0.0                   0.0   
1                    0.0                   0.0                   0.0   
2                    0.0                   0.0                   0.0   
3                    0.0                   0.0                   0.0   
4                    0.0                   0.0                   0.0   

  segmentsCabinCode_Leg1 segmentsCabinCode_Leg2  segmentsCabinCode_Leg3  \
0                  coach                      0                       0   
1                  coach                      0                       0   
2                  coach                      0                       0   
3                  coach                      0                       0   
4                  coach                      0                       0   

   segmentsCabinCode_Leg4  traveltime_hours  
0                       0          2.483333  
1                       0          2.500000  
2                       0          2.500000  
3                       0          2.533333  
4                       0          2.566667  

[5 rows x 58 columns]

In [3]:
df.tail()

searchDate  flightDate startingAirport destinationAirport  \
13519994  2022-05-19  2022-06-03             SFO                PHL   
13519995  2022-05-19  2022-06-03             SFO                PHL   
13519996  2022-05-19  2022-06-03             SFO                PHL   
13519997  2022-05-19  2022-06-03             SFO                PHL   
13519998  2022-05-19  2022-06-03             SFO                PHL   

          isBasicEconomy  isRefundable  isNonStop    totalFare  \
13519994           False         False      False   836.599976   
13519995           False         False      False   837.599976   
13519996           False         False      False   837.599976   
13519997           False         False      False   837.599976   
13519998           False         False       True  1649.599976   

          totalTravelDistance  segmentsDepartureTimeEpochSeconds_Leg1  ...  \
13519994                  0.0                            1.654261e+09  ...   
13519995                  0.0                            1.654258e+09  ...   
13519996                  0.0                            1.654287e+09  ...   
13519997                  0.0                            1.654267e+09  ...   
13519998                  0.0                            1.654286e+09  ...   

          segmentsDurationInSeconds_Leg4  segmentsDistance_Leg1  \
13519994                             0.0                    0.0   
13519995                             0.0                    0.0   
13519996                             0.0                    0.0   
13519997                             0.0                    0.0   
13519998                             0.0                    0.0   

          segmentsDistance_Leg2 segmentsDistance_Leg3 segmentsDistance_Leg4  \
13519994                    0.0                   0.0                   0.0   
13519995                    0.0                   0.0                   0.0   
13519996                    0.0                   0.0                   0.0   
13519997                    0.0                   0.0                   0.0   
13519998                    0.0                   0.0                   0.0   

         segmentsCabinCode_Leg1 segmentsCabinCode_Leg2  \
13519994                  coach                  coach   
13519995                  coach                  coach   
13519996                  coach                  coach   
13519997                  coach                  coach   
13519998                  coach                      0   

          segmentsCabinCode_Leg3  segmentsCabinCode_Leg4  traveltime_hours  
13519994                       0                       0          8.283333  
13519995                       0                       0          7.333333  
13519996                       0                       0          7.816667  
13519997                       0                       0          9.166667  
13519998                       0                       0          5.433333  

[5 rows x 58 columns]

In [4]:
df.shape

(13519999, 58)

In [5]:
df.describe()

totalFare  totalTravelDistance  \
count  1.352000e+07         1.352000e+07   
mean   3.752204e+02         1.438146e+03   
std    2.058691e+02         8.993270e+02   
min    2.397000e+01         0.000000e+00   
25%    2.235800e+02         7.330000e+02   
50%    3.446000e+02         1.321000e+03   
75%    4.876000e+02         2.305000e+03   
max    8.260610e+03         4.430000e+03   

       segmentsDepartureTimeEpochSeconds_Leg1  \
count                            1.352000e+07   
mean                             1.928912e+09   
std                              2.692376e+08   
min                              1.650177e+09   
25%                              1.652514e+09   
50%                              1.653666e+09   
75%                              1.655074e+09   
max                              1.658192e+09   

       segmentsDepartureTimeEpochSeconds_Leg2  \
count                            1.352000e+07   
mean                             1.300563e+09   
std                              7.278751e+08   
min                              0.000000e+00   
25%                              0.000000e+00   
50%                              1.652725e+09   
75%                              1.654449e+09   
max                              1.658228e+09   

       segmentsDepartureTimeEpochSeconds_Leg3  \
count                            1.352000e+07   
mean                             1.211704e+08   
std                              4.177648e+08   
min                              0.000000e+00   
25%                              0.000000e+00   
50%                              0.000000e+00   
75%                              0.000000e+00   
max                              1.658185e+09   

       segmentsDepartureTimeEpochSeconds_Leg4  \
count                            1.352000e+07   
mean                             4.614340e+06   
std                              8.711446e+07   
min                              0.000000e+00   
25%                              0.000000e+00   
50%                              0.000000e+00   
75%                              0.000000e+00   
max                              1.658159e+09   

       segmentsArrivalTimeEpochSeconds_Leg1  \
count                          1.352000e+07   
mean                           1.928912e+09   
std                            2.692340e+08   
min                            1.650187e+09   
25%                            1.652527e+09   
50%                            1.653676e+09   
75%                            1.655084e+09   
max                            1.658203e+09   

       segmentsArrivalTimeEpochSeconds_Leg2  \
count                          1.352000e+07   
mean                           1.300564e+09   
std                            7.278762e+08   
min                            0.000000e+00   
25%                            0.000000e+00   
50%                            1.652735e+09   
75%                            1.654458e+09   
max                            1.658241e+09   

       segmentsArrivalTimeEpochSeconds_Leg3  \
count                          1.352000e+07   
mean                           1.211705e+08   
std                            4.177653e+08   
min                            0.000000e+00   
25%                            0.000000e+00   
50%                            0.000000e+00   
75%                            0.000000e+00   
max                            1.658191e+09   

       segmentsArrivalTimeEpochSeconds_Leg4  segmentsDurationInSeconds_Leg1  \
count                          1.352000e+07                    1.352000e+07   
mean                           4.614386e+06                    9.756050e+03   
std                            8.711457e+07                    4.501104e+03   
min                            0.000000e+00                    2.100000e+03   
25%                            0.000000e+00                    6.180000e+03   
50%                            0.000000e+00                    8.280000e+03   
75%                 

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13519999 entries, 0 to 13519998
Data columns (total 58 columns):
 #   Column                                  Dtype   
---  ------                                  -----   
 0   searchDate                              object  
 1   flightDate                              object  
 2   startingAirport                         category
 3   destinationAirport                      category
 4   isBasicEconomy                          bool    
 5   isRefundable                            bool    
 6   isNonStop                               bool    
 7   totalFare                               float32 
 8   totalTravelDistance                     float32 
 9   segmentsDepartureTimeEpochSeconds_Leg1  float32 
 10  segmentsDepartureTimeEpochSeconds_Leg2  float32 
 11  segmentsDepartureTimeEpochSeconds_Leg3  float32 
 12  segmentsDepartureTimeEpochSeconds_Leg4  float32 
 13  segmentsDepartureTimeRaw_Leg1           object  
 14  segmentsDepartur

In [7]:
data = df.copy()

# 2. Data Understanding 

### a. Checking for null values

In [8]:
data.isnull().sum()

searchDate                                     0
flightDate                                     0
startingAirport                                0
destinationAirport                             0
isBasicEconomy                                 0
isRefundable                                   0
isNonStop                                      0
totalFare                                      0
totalTravelDistance                            0
segmentsDepartureTimeEpochSeconds_Leg1         0
segmentsDepartureTimeEpochSeconds_Leg2         0
segmentsDepartureTimeEpochSeconds_Leg3         0
segmentsDepartureTimeEpochSeconds_Leg4         0
segmentsDepartureTimeRaw_Leg1                  0
segmentsDepartureTimeRaw_Leg2                  0
segmentsDepartureTimeRaw_Leg3                  0
segmentsDepartureTimeRaw_Leg4                  0
segmentsArrivalTimeEpochSeconds_Leg1           0
segmentsArrivalTimeEpochSeconds_Leg2           0
segmentsArrivalTimeEpochSeconds_Leg3           0
segmentsArrivalTimeE

### b. Checking for duplicates

In [9]:
data.duplicated().sum()

0

### c. Unique values and counts for columns

##### For 'category' type columns

In [10]:
# Get unique values and count for each 'category' column
category_columns_unique_values = {col: data[col].unique() for col in data.select_dtypes(['category']).columns}
category_columns_unique_counts = {col: data[col].nunique() for col in data.select_dtypes(['category']).columns}

- Unique Values

In [11]:
# Print the results
print("Unique values in each 'category' column:")
for col, values in category_columns_unique_values.items():
    print(f"{col}: {values}")

Unique values in each 'category' column:
startingAirport: ['ATL', 'BOS', 'CLT', 'DEN', 'DFW', ..., 'MIA', 'OAK', 'ORD', 'PHL', 'SFO']
Length: 16
Categories (16, object): ['ATL', 'BOS', 'CLT', 'DEN', ..., 'OAK', 'ORD', 'PHL', 'SFO']
destinationAirport: ['BOS', 'CLT', 'DEN', 'DFW', 'DTW', ..., 'OAK', 'ORD', 'PHL', 'SFO', 'ATL']
Length: 16
Categories (16, object): ['ATL', 'BOS', 'CLT', 'DEN', ..., 'OAK', 'ORD', 'PHL', 'SFO']
segmentsArrivalAirportCode_Leg1: ['BOS', 'CLT', 'IAD', 'ORD', 'FLL', ..., 'SUN', 'PSC', 'BUR', 'BFL', 'KOA']
Length: 199
Categories (199, object): ['ABE', 'ABQ', 'ACK', 'ACV', ..., 'TVC', 'TYS', 'UIN', 'XNA']
segmentsArrivalAirportCode_Leg2: [0, 'BOS', 'CLT', 'DEN', 'DFW', ..., 'GRB', 'LBB', 'IPL', 'SRQ', 'TRI']
Length: 181
Categories (181, object): [0, 'ABQ', 'ACK', 'ACV', ..., 'UIN', 'VPS', 'XNA', 'YUM']
segmentsArrivalAirportCode_Leg3: [0, 'STL', 'IAD', 'BNA', 'JFK', ..., 'ACK', 'MVY', 'IAH', 'DCA', 'MSP']
Length: 31
Categories (31, object): [0, 'ACK', 'ATL', 'AUS'

- Counts

In [12]:
print("\nCount of unique values in each 'category' column:")
for col, count in category_columns_unique_counts.items():
    print(f"{col}: {count}")


Count of unique values in each 'category' column:
startingAirport: 16
destinationAirport: 16
segmentsArrivalAirportCode_Leg1: 199
segmentsArrivalAirportCode_Leg2: 181
segmentsArrivalAirportCode_Leg3: 31
segmentsArrivalAirportCode_Leg4: 18
segmentsDepartureAirportCode_Leg1: 23
segmentsDepartureAirportCode_Leg2: 198
segmentsDepartureAirportCode_Leg3: 178
segmentsDepartureAirportCode_Leg4: 24
segmentsAirlineName_Leg1: 14
segmentsAirlineName_Leg2: 15
segmentsAirlineName_Leg3: 13
segmentsAirlineName_Leg4: 7
segmentsAirlineCode_Leg1: 14
segmentsAirlineCode_Leg2: 15
segmentsAirlineCode_Leg3: 13
segmentsAirlineCode_Leg4: 7
segmentsEquipmentDescription_Leg1: 43
segmentsEquipmentDescription_Leg2: 43
segmentsEquipmentDescription_Leg3: 44
segmentsEquipmentDescription_Leg4: 23
segmentsCabinCode_Leg1: 4
segmentsCabinCode_Leg2: 5
segmentsCabinCode_Leg3: 5
segmentsCabinCode_Leg4: 4


##### For 'object' type columns

In [14]:
# Get unique values and count for each 'object' column
object_columns_unique_values = {col: data[col].unique() for col in data.select_dtypes(['object']).columns}
object_columns_unique_counts = {col: data[col].nunique() for col in data.select_dtypes(['object']).columns}

- Unique Values

In [15]:
print("\nUnique values in each 'object' column:")
for col, values in object_columns_unique_values.items():
    print(f"{col}: {values}")


Unique values in each 'object' column:
searchDate: ['2022-04-16' '2022-04-17' '2022-04-18' '2022-04-19' '2022-04-20'
 '2022-04-21' '2022-04-22' '2022-04-23' '2022-04-24' '2022-04-25'
 '2022-04-26' '2022-04-27' '2022-04-28' '2022-04-29' '2022-04-30'
 '2022-05-01' '2022-05-02' '2022-05-03' '2022-05-04' '2022-05-05'
 '2022-05-06' '2022-05-07' '2022-05-08' '2022-05-09' '2022-05-10'
 '2022-05-11' '2022-05-12' '2022-05-13' '2022-05-14' '2022-05-17'
 '2022-05-18' '2022-05-19']
flightDate: ['2022-04-17' '2022-04-18' '2022-04-19' '2022-04-20' '2022-04-21'
 '2022-04-22' '2022-04-23' '2022-04-24' '2022-04-25' '2022-04-26'
 '2022-04-27' '2022-04-28' '2022-04-29' '2022-04-30' '2022-05-01'
 '2022-05-02' '2022-05-03' '2022-05-04' '2022-05-05' '2022-05-06'
 '2022-05-07' '2022-05-08' '2022-05-09' '2022-05-10' '2022-05-11'
 '2022-05-12' '2022-05-13' '2022-05-14' '2022-05-15' '2022-05-16'
 '2022-05-17' '2022-05-18' '2022-05-19' '2022-05-20' '2022-05-21'
 '2022-05-22' '2022-05-23' '2022-05-24' '2022-05-2

- Counts

In [16]:
print("\nCount of unique values in each 'object' column:")
for col, count in object_columns_unique_counts.items():
    print(f"{col}: {count}")


Count of unique values in each 'object' column:
searchDate: 32
flightDate: 92
segmentsDepartureTimeRaw_Leg1: 144495
segmentsDepartureTimeRaw_Leg2: 144672
segmentsDepartureTimeRaw_Leg3: 50180
segmentsDepartureTimeRaw_Leg4: 2245
segmentsArrivalTimeRaw_Leg1: 194851
segmentsArrivalTimeRaw_Leg2: 181340
segmentsArrivalTimeRaw_Leg3: 53865
segmentsArrivalTimeRaw_Leg4: 2325


##### For 'bool' type columns

In [17]:
# Get unique values and count for each 'bool' column
object_columns_unique_values = {col: data[col].unique() for col in data.select_dtypes(['bool']).columns}
object_columns_unique_counts = {col: data[col].nunique() for col in data.select_dtypes(['bool']).columns}

- Unique Values

In [18]:
print("\nUnique values in each 'bool' column:")
for col, values in object_columns_unique_values.items():
    print(f"{col}: {values}")


Unique values in each 'bool' column:
isBasicEconomy: [False  True]
isRefundable: [False  True]
isNonStop: [ True False]


- Counts

In [19]:
print("\nCount of unique values in each 'bool' column:")
for col, count in object_columns_unique_counts.items():
    print(f"{col}: {count}")


Count of unique values in each 'bool' column:
isBasicEconomy: 2
isRefundable: 2
isNonStop: 2


# 3. Data Preparation and Feature Engineering

- During the merging the missing values for columns were replaced by 0 and hence, this codes checks on the percentage of missing values in each column

In [20]:
def calculate_zeros_all_types(df):
    zero_count = {}
    percent_zeros = {}

    for column in df.columns:
        # Ensure the column is numeric before comparing to zero
        if pd.api.types.is_numeric_dtype(df[column]):
            zero_count[column] = (df[column] == 0).sum()
            percent_zeros[column] = (zero_count[column] / len(df)) * 100
        elif pd.api.types.is_string_dtype(df[column]) or pd.api.types.is_categorical_dtype(df[column]):
            # Count occurrences of '0' as a string for non-numeric data
            zero_count[column] = (df[column].astype(str) == '0').sum()
            percent_zeros[column] = (zero_count[column] / len(df)) * 100
        else:
            # For non-numeric and non-categorical data, we can't compare with zero
            zero_count[column] = np.nan
            percent_zeros[column] = np.nan

    return pd.DataFrame({'Count of 0': zero_count, 'Percentage of 0': percent_zeros})

In [21]:
calculate_zeros_all_types(data)

Count of 0  Percentage of 0
searchDate                                       0         0.000000
flightDate                                       0         0.000000
startingAirport                                  0         0.000000
destinationAirport                               0         0.000000
isBasicEconomy                            12768661        94.442766
isRefundable                              13519808        99.998587
isNonStop                                  9564064        70.740124
totalFare                                        0         0.000000
totalTravelDistance                         959619         7.097774
segmentsDepartureTimeEpochSeconds_Leg1           0         0.000000
segmentsDepartureTimeEpochSeconds_Leg2     3955935        29.259876
segmentsDepartureTimeEpochSeconds_Leg3    12525106        92.641323
segmentsDepartureTimeEpochSeconds_Leg4    13482281        99.721021
segmentsDepartureTimeRaw_Leg1                    0         0.000000
segmentsDepartureTimeRaw_Leg2              3955935        29.259876
segmentsDepartureTimeRaw_Leg3             12525106        92.641323
segmentsDepartureTimeRaw_Leg4             13482281        99.721021
segmentsArrivalTimeEpochSeconds_Leg1             0         0.000000
segmentsArrivalTimeEpochSeconds_Leg2       3955935        29.259876
segmentsArrivalTimeEpochSeconds_Leg3      12525106        92.641323
segmentsArrivalTimeEpochSeconds_Leg4      13482281        99.721021
segmentsArrivalTimeRaw_Leg1                      0         0.000000
segmentsArrivalTimeRaw_Leg2                3955935        29.259876
segmentsArrivalTimeRaw_Leg3               12525106        92.641323
segmentsArrivalTimeRaw_Leg4               13482281        99.721021
segmentsArrivalAirportCode_Leg1                  0         0.000000
segmentsArrivalAirportCode_Leg2            3955935        29.259876
segmentsArrivalAirportCode_Leg3           12525106        92.641323
segmentsArrivalAirportCode_Leg4           13482281        99.721021
segmentsDepartureAirportCode_Leg1                0         0.000000
segmentsDepartureAirportCode_Leg2          3955935        29.259876
segmentsDepartureAirportCode_Leg3         12525106        92.641323
segmentsDepartureAirportCode_Leg4         13482281        99.721021
segmentsAirlineName_Leg1                         0         0.000000
segmentsAirlineName_Leg2                   3955935        29.259876
segmentsAirlineName_Leg3                  12525106        92.641323
segmentsAirlineName_Leg4                  13482281        99.721021
segmentsAirlineCode_Leg1                         0         0.000000
segmentsAirlineCode_Leg2                   3955935        29.259876
segmentsAirlineCode_Leg3                  12525106        92.641323
segmentsAirlineCode_Leg4                  13482281        99.721021
segmentsEquipmentDescription_Leg1           262676         1.942870
segmentsEquipmentDescription_Leg2          3955935        29.259876
segmentsEquipmentDescription_Leg3         12525106        92.641323
segmentsEquipmentDescription_Leg4         13482281        99.721021
segmentsDurationInSeconds_Leg1                   0         0.000000
segmentsDurationInSeconds_Leg2             3955935        29.259876
segmentsDurationInSeconds_Leg3            12525106        92.641323
segmentsDurationInSeconds_Leg4            13482281        99.721021
segmentsDistance_Leg1                       959619         7.097774
segmentsDistance_Leg2                      4788569        35.418412
segmentsDistance_Leg3                     12552974        92.847448
segmentsDistance_Leg4                     13482281        99.721021
segmentsCabinCode_Leg1                           0         0.000000
segmentsCabinCode_Leg2                     3955935        29.259876
segmentsCabinCode_Leg3                    12525106        92.641323
segmentsCabinCode_Leg4                    13482281        99.721021
traveltime_hours                                 0         0.000000

In [22]:
data['segmentsCabinCode_Leg1'].value_counts()

coach            13472653
premium coach       22826
first               21120
business             3400
Name: segmentsCabinCode_Leg1, dtype: int64

In [24]:
data['segmentsCabinCode_Leg2'].value_counts()

coach            9538989
0                3955935
first              13174
premium coach      10672
business            1229
Name: segmentsCabinCode_Leg2, dtype: int64

- Converting the column 'flightDate' to datetime to extract more features of the fligth day, week and month.  

In [25]:
data['flightDate'] = pd.to_datetime(data['flightDate'])

In [26]:
data['flightMonth'] = data['flightDate'].dt.month
data['flightDay'] = data['flightDate'].dt.day
data['flightDayOfWeek'] = data['flightDate'].dt.dayofweek
data['flightWeekOfYear'] = data['flightDate'].dt.isocalendar().week
data['flightIsWeekend'] = data['flightDate'].dt.dayofweek >= 5

- Dropping columns related to Leg 2, Leg 3 and Leg 4 as they are not required for futher modeling purposes. 

In [27]:
columns_to_drop = [col for col in data.columns if 'Leg2' in col or 'Leg3' in col or 'Leg4' in col]
data = data.drop(columns=columns_to_drop)

In [28]:
additional_columns_to_drop = ['searchDate', 'isBasicEconomy', 'isRefundable', 'isNonStop', 'segmentsAirlineName_Leg1', 'segmentsAirlineCode_Leg1', 'segmentsEquipmentDescription_Leg1', 'segmentsDurationInSeconds_Leg1']
data = data.drop(columns=additional_columns_to_drop)

- Creating a new column cabin type to extract the type and save as a different column.

In [29]:
data['cabin_type'] = data['segmentsCabinCode_Leg1']
data = data.drop(columns=['segmentsCabinCode_Leg1'])

- The rows which have missing values in 'traveltime_hours' and 'totalTravelDistance' have been dropped as their percentage is significantly low and there is still a lot of data for modeling purposes.

In [30]:
data = data.dropna(subset=['traveltime_hours'])

In [31]:
data = data[data['totalTravelDistance'] != 0]

- Converting the column 'segmentsDepartureTimeRaw_Leg1' to datetime to extract a new column named 'departure time'.

In [32]:
data['segmentsDepartureTimeRaw_Leg1'] = pd.to_datetime(data['segmentsDepartureTimeRaw_Leg1'], utc=True)

In [33]:
print(data['segmentsDepartureTimeRaw_Leg1'].dtype)

datetime64[ns, UTC]


In [34]:
# Defining a function to categorize the time of day
def categorize_time_of_day(hour):
    if 5 <= hour < 12:
        return 'morning'
    elif 12 <= hour < 17:
        return 'afternoon'
    elif 17 <= hour < 21:
        return 'evening'
    else:
        return 'night'

In [35]:
data['departure_time'] = data['segmentsDepartureTimeRaw_Leg1'].dt.hour.apply(categorize_time_of_day)

In [36]:
data[['segmentsDepartureTimeRaw_Leg1', 'departure_time']]

segmentsDepartureTimeRaw_Leg1 departure_time
0            2022-04-17 16:57:00+00:00      afternoon
1            2022-04-17 10:30:00+00:00        morning
2            2022-04-17 15:35:00+00:00      afternoon
3            2022-04-17 17:59:00+00:00        evening
4            2022-04-17 13:59:00+00:00      afternoon
...                                ...            ...
13519914     2022-06-03 12:30:00+00:00      afternoon
13519915     2022-06-03 14:00:00+00:00      afternoon
13519916     2022-06-03 23:25:00+00:00          night
13519917     2022-06-03 20:35:00+00:00        evening
13519918     2022-06-03 20:06:00+00:00        evening

[12285686 rows x 2 columns]

- Dropping the last chunk of unneccessary columns 

In [37]:
columns_to_check_and_drop = [
    'segmentsDepartureTimeEpochSeconds_Leg1',
    'flightDate',
    'segmentsDepartureTimeRaw_Leg1',
    'segmentsArrivalTimeEpochSeconds_Leg1',
    'segmentsArrivalTimeRaw_Leg1',
    'segmentsArrivalAirportCode_Leg1',
    'segmentsDepartureAirportCode_Leg1',
    'segmentsDistance_Leg1'
]

columns_to_drop_1 = [col for col in columns_to_check_and_drop if col in data.columns]
data = data.drop(columns=columns_to_drop_1)

- Final data checks before modeling 

In [38]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12285686 entries, 0 to 13519918
Data columns (total 12 columns):
 #   Column               Dtype   
---  ------               -----   
 0   startingAirport      category
 1   destinationAirport   category
 2   totalFare            float32 
 3   totalTravelDistance  float32 
 4   traveltime_hours     float64 
 5   flightMonth          int64   
 6   flightDay            int64   
 7   flightDayOfWeek      int64   
 8   flightWeekOfYear     UInt32  
 9   flightIsWeekend      bool    
 10  cabin_type           category
 11  departure_time       object  
dtypes: UInt32(1), bool(1), category(3), float32(2), float64(1), int64(3), object(1)
memory usage: 761.6+ MB


In [39]:
final_data = data.copy()

In [40]:
final_data.head()

startingAirport destinationAirport   totalFare  totalTravelDistance  \
0             ATL                BOS  248.600006                947.0   
1             ATL                BOS  248.600006                947.0   
2             ATL                BOS  248.600006                947.0   
3             ATL                BOS  248.600006                947.0   
4             ATL                BOS  248.600006                947.0   

   traveltime_hours  flightMonth  flightDay  flightDayOfWeek  \
0          2.483333            4         17                6   
1          2.500000            4         17                6   
2          2.500000            4         17                6   
3          2.533333            4         17                6   
4          2.566667            4         17                6   

   flightWeekOfYear  flightIsWeekend cabin_type departure_time  
0                15             True      coach      afternoon  
1                15             True      coach        morning  
2                15             True      coach      afternoon  
3                15             True      coach        evening  
4                15             True      coach      afternoon

# 4. Building Data Pipeline for Modeling

In [41]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.dummy import DummyRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error

In [42]:
# Define the features and the target
X = final_data.drop('totalFare', axis=1)
y = final_data['totalFare']

##### Creating Train, Test and Validation Sets

In [43]:
# Split the data into train, validation, and test sets
X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.4, random_state=42)
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)

##### One-Hot Encoding the Categorical Columns

In [44]:
# Define categorical columns for one-hot encoding
categorical_cols = X.select_dtypes(include=['category', 'object', 'bool']).columns

In [45]:
# Create a ColumnTransformer for one-hot encoding categorical variables
preprocessor = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_cols)
    ],
    remainder='passthrough'  # Leave other columns untouched
)

##### Creating a Baseline model and Gradient Boost model.

In [46]:
# Define the baseline model (DummyRegressor) and Gradient Boosting Regressor
baseline_model = DummyRegressor(strategy='mean')
gradient_boosting_model = GradientBoostingRegressor(random_state=42)

In [47]:
# Create pipelines for both models
baseline_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                                    ('model', baseline_model)])
gradient_boosting_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                                             ('model', gradient_boosting_model)])

- The model training time is roughly 30-45 mins

In [48]:
# Fit the pipelines to the training data
baseline_pipeline.fit(X_train, y_train)
gradient_boosting_pipeline.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('cat',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  Index(['startingAirport', 'destinationAirport', 'flightIsWeekend',
       'cabin_type', 'departure_time'],
      dtype='object'))])),
                ('model', GradientBoostingRegressor(random_state=42))])

In [49]:
# Predict on the validation set
y_val_pred_baseline = baseline_pipeline.predict(X_val)
y_val_pred_gb = gradient_boosting_pipeline.predict(X_val)

##### Calculating the model scores using RMSE, MSE and MAE

In [50]:
# Calculate metrics for baseline
mse_baseline = mean_squared_error(y_val, y_val_pred_baseline)
rmse_baseline = mean_squared_error(y_val, y_val_pred_baseline, squared=False)
mae_baseline = mean_absolute_error(y_val, y_val_pred_baseline)

# Calculate metrics for gradient boosting
mse_gb = mean_squared_error(y_val, y_val_pred_gb)
rmse_gb = mean_squared_error(y_val, y_val_pred_gb, squared=False)
mae_gb = mean_absolute_error(y_val, y_val_pred_gb)

# Display metrics for both models
print(f"Baseline Model: MSE = {mse_baseline}, RMSE = {rmse_baseline}, MAE = {mae_baseline}")
print(f"Gradient Boosting Model: MSE = {mse_gb}, RMSE = {rmse_gb}, MAE = {mae_gb}")

Baseline Model: MSE = 44166.2421875, RMSE = 210.15765380859375, MAE = 156.8333282470703
Gradient Boosting Model: MSE = 19159.73911823398, RMSE = 138.4187094226571, MAE = 95.99653903259082


In [52]:
from joblib import dump

# Save the Gradient Boosting model pipeline to a joblib file
dump(gradient_boosting_pipeline, '../../models/Varun/gradient_boosting_model.joblib')

['../../models/Varun/gradient_boosting_model.joblib']

### Model Performance Metrics

- 1. Mean Squared Error (MSE): The average squared difference between the estimated values and the actual value. It gives a general idea of the magnitude of error but is sensitive to outliers since the errors are squared.
- 2. Root Mean Squared Error (RMSE): The square root of the MSE. It's in the same units as the target variable, making it more interpretable. 
- 3. Mean Absolute Error (MAE): The average absolute difference between the predicted values and the actual values. It provides a linear score, meaning all individual differences are weighted equally in the average.

### Model Result Analysis

* Comparison to Baseline: Both the MSE and RMSE are lower for the Gradient Boosting Model compared to the Baseline Model, which means that the Gradient Boosting Model has a better fit to the data and makes more accurate predictions. The MAE is also lower, indicating that on average, the Gradient Boosting Model's predictions are closer to the actual values.

* Relative Improvement: The Gradient Boosting Model has reduced the MSE to almost half of what the Baseline Model's MSE is, and similar improvement is seen in RMSE and MAE. This is a substantial improvement in predictive performance.

* Error Magnitude: The RMSE of the Gradient Boosting Model is 138.42, which means that on average, the model's predictions are off by this amount from the actual fare prices. This error may or may not be acceptable. For high-cost items (like expensive flights), this might be a reasonable error. For low-cost items, this might be too high.

* Model Selection: Given that the Gradient Boosting Model significantly outperforms the Baseline Model in all three metrics, it is likely a better model for your data and can be selected for further optimization and testing.

* In summary, the Gradient Boosting Model shows a significant improvement over the Baseline Model, indicating that it is learning from the features provided and is able to make more accurate predictions. Further model tuning, feature engineering, possibly doing GridSearchCV or model stacking could lead to even better results. It is also important to note that the datset is huge and training models takes significant amount of time and hence futher tuning must be done where computational powers are high or the modeling is done in a small subset of data.

In [95]:
# Example DataFrame with the same structure as your training data
X_sample_6= pd.DataFrame({
    'startingAirport': ['JFK'],
    'destinationAirport': ['LAX'],
    'totalTravelDistance': [2475.0],
    'traveltime_hours': [5.5],
    'flightMonth': [1],
    'flightDay': [15],
    'flightDayOfWeek': [2],
    'flightWeekOfYear': [3],
    'flightIsWeekend': [False],
    'cabin_type': ['premium coach'],
    'departure_time': ['morning']
})

In [96]:
predictions_6 = gradient_boosting_pipeline.predict(X_sample_6)

In [97]:
print(predictions_6)

[797.93965266]


In [75]:
# Filter the DataFrame
filtered_df = data[
    (data['startingAirport'] == 'JFK') & 
    (data['destinationAirport'] == 'LAX') &
    (data['cabin_type'] == 'first')
]


In [76]:
if not filtered_df.empty:
    print("There are records matching the criteria.")
    print(filtered_df)
else:
    print("No records match the criteria.")

There are records matching the criteria.
        startingAirport destinationAirport    totalFare  totalTravelDistance  \
6426585             JFK                LAX  2652.600098               2458.0   
6426886             JFK                LAX  3810.600098               2458.0   
6430673             JFK                LAX  3810.600098               2458.0   
6706362             JFK                LAX  4752.600098               2458.0   
6788533             JFK                LAX  4752.600098               2458.0   
6851581             JFK                LAX  2658.610107               2458.0   
6851582             JFK                LAX  2658.610107               2458.0   
6876859             JFK                LAX  2658.610107               2458.0   
6903409             JFK                LAX  2658.610107               2458.0   
6929807             JFK                LAX  2658.610107               2458.0   
6929808             JFK                LAX  2658.610107               2458.0   